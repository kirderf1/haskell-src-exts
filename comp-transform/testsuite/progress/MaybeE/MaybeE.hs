{-# LANGUAGE ComposableTypes #-}

piececategory A

data piece A ==> B = B Int
data piece A ==> C = C
data piece A ==> D = D A

type E = A ==> (B, C, D)

e :: Int -> Maybe E
e = undefined

main = putStrLn $ "Det funkar!"
