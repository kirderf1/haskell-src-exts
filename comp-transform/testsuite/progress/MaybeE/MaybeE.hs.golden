{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FlexibleContexts #-}
import qualified Data.Comp
import qualified Data.Comp.Derive
import qualified Data.Comp.Show ()
import qualified Data.Comp.Equality ()

data B a = B Int
             deriving Functor

$(
  Data.Comp.Derive.derive
    [Data.Comp.Derive.makeTraversable, Data.Comp.Derive.makeFoldable,
     Data.Comp.Derive.makeEqF, Data.Comp.Derive.makeShowF,
     Data.Comp.Derive.smartConstructors,
     Data.Comp.Derive.smartAConstructors]
    [''B]
  )

data C a = C deriving Functor

$(
  Data.Comp.Derive.derive
    [Data.Comp.Derive.makeTraversable, Data.Comp.Derive.makeFoldable,
     Data.Comp.Derive.makeEqF, Data.Comp.Derive.makeShowF,
     Data.Comp.Derive.smartConstructors,
     Data.Comp.Derive.smartAConstructors]
    [''C]
  )

data D a = D a
             deriving Functor

$(
  Data.Comp.Derive.derive
    [Data.Comp.Derive.makeTraversable, Data.Comp.Derive.makeFoldable,
     Data.Comp.Derive.makeEqF, Data.Comp.Derive.makeShowF,
     Data.Comp.Derive.smartConstructors,
     Data.Comp.Derive.smartAConstructors]
    [''D]
  )

type E = Data.Comp.Term (B Data.Comp.:+: C Data.Comp.:+: D)

e :: Int -> Maybe E
e = undefined
main = putStrLn $ "Det funkar!"
